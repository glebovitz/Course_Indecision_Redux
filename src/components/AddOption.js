import React from 'react';

export default class AddOption extends React.Component {
    state = {
        inputValue: '',
        optionCount: 0 //used to store previous optionCount prop
    };

    // used changes in optionCount to determine that we successfully added
    // a new option. Previous optionCount is stored in local state.
    static getDerivedStateFromProps = (nextProps, prevState) => {
        if (nextProps.optionCount !== prevState.optionCount) {
            return { 
                inputValue: '',
                optionCount: nextProps.optionCount,
            };
        } else {
            return null;
        }
    };
    handleAddOption = (e) => {
        e.preventDefault();
        this.props.addOption(this.state.inputValue);
    };
    handleOnChange = (e) => {
        const inputValue = e.target.value;
        this.setState(() => ({
            inputValue
        }));
    };
  render() {
      return (
        <div>
            {!!this.props.error && <p className='add-option-error'>{this.props.error}</p>}
            <form className='add-option' onSubmit={this.handleAddOption}>
                <input 
                    className='add-option__input'
                    type='text'
                    value={this.state.inputValue}
                    onChange={this.handleOnChange}
                />
                <button className='button'>Add Option</button>
            </form>
        </div>
      );
  }
}
