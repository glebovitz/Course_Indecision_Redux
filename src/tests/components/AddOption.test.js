import React from 'react';
import { shallow } from 'enzyme';
import AddOption from '../../components/AddOption';

test('should render Component correctly', () => {
  const wrapper = shallow(<AddOption error={undefined} optionCount={1}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render Component correctly', () => {
  const wrapper = shallow(<AddOption error={'testabcde'} optionCount={0}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should set inputValue on input', () => {
  const value = 1;
  const wrapper = shallow(<AddOption error={undefined} optionCount={value}/>);
  expect(wrapper.state('optionCount')).toBe(value);
});

test('should set inputValue on input', () => {
  const value = 'abc123';
  const wrapper = shallow(<AddOption error={undefined} optionCount={1}/>);
  wrapper.find('input').simulate('change', {
    target: { value }
  });
  expect(wrapper.state('inputValue')).toBe(value);
});

test('should clear state inputValue on optionCount change', () => {
  const value = 'abc123';
  const wrapper = shallow(<AddOption error={undefined} optionCount={0}/>);
  wrapper.find('input').simulate('change', { target: { value }});
  wrapper.setProps({ optionCount: 1});
  expect(wrapper.state('inputValue')).toBe('');
});

test('should call handleAddOption with inputValue on submit', () => {
  const value = 'abc123';
  const addOption = jest.fn();
  const wrapper = shallow (
    <AddOption 
      error={undefined} 
      optionCount={0}
      addOption={addOption}
    />
  );
  wrapper.find('input').simulate('change', { target: { value }});
  wrapper.find('form').simulate("submit", { preventDefault: () => {} });
  expect(addOption).toHaveBeenLastCalledWith(value);
});

