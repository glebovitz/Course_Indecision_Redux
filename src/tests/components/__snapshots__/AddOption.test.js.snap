// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Component correctly 1`] = `
<div>
  <form
    className="add-option"
    onSubmit={[Function]}
  >
    <input
      className="add-option__input"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      className="button"
    >
      Add Option
    </button>
  </form>
</div>
`;

exports[`should render Component correctly 2`] = `
<div>
  <p
    className="add-option-error"
  >
    testabcde
  </p>
  <form
    className="add-option"
    onSubmit={[Function]}
  >
    <input
      className="add-option__input"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      className="button"
    >
      Add Option
    </button>
  </form>
</div>
`;
